### PR TITLE
Add BLOOM config with smart attention caches, fix minor issues

### PR DESCRIPTION
**Pull request by justheuristic**
This pull request

- adds a (highly tuned) bloom config that stores attention caches on separate devices instead of just device 0
   - this is important for inference with long sequences, where passing all attention cache from device 0 would be longer than recomputing it
- fixes several issues that occur with the odd number of gpus (e.g. 3)

### CODE REVIEW

File: src/tensor_parallel/communications.py [hunk 1]
```diff
@@ -58,16 +58,45 @@ def __init__(self, world_size: int):
...
         self.dim = dim
 
-    def _nccl_allgather(self, *tensors: torch.Tensor):
-        gathered_tensors = cross_device_ops.NCCLAllGatherFunction.apply(*tensors)
-        dim_indices = list(range(1, gathered_tensors[0].ndim))
-        dim_indices.insert(self.dim, 0)
-        concatenated_shape = list(tensors[0].shape)
-        concatenated_shape[self.dim] = -1
-        return tuple(output.permute(dim_indices).reshape(concatenated_shape) for output in gathered_tensors)
+    def __call__(self, x: torch.Tensor, rank: int):
```
Comment (justheuristic) [inline src/tensor_parallel/communications.py:65]: NCCL's allgather requires tensors to have the same rank; this code circumvents the problem by padding tensors before allgather and un-padding afterwards

File: src/tensor_parallel/slicer_wrapper.py [hunk 2]
```diff
@@ -221,7 +236,7 @@ def process_state_(module: nn.Module, config: Config, *, rank: int, world_size:
                 unused_patterns.discard(pattern)
 
     if unused_patterns:
-        logger.warning(f"The following patterns in state_rules were unused: {sorted(unused_patterns)}")
+        logger.warning(f"The following patterns in state_rules were unused: {[str(p) for p in unused_patterns]}")
```
Comment (justheuristic) [inline src/tensor_parallel/slicer_wrapper.py:239]: this would fail because you cannot sort regex patterns

File: src/tensor_parallel/slicer_wrapper.py [hunk 3]
```diff
@@ -247,9 +260,12 @@ def process_output(
...
 ):
     if isinstance(output, torch.Tensor):
-        return process_output([output], output_actions, rank=rank, world_size=world_size)[0]
+        return process_output({0: output}, output_actions, rank=rank, world_size=world_size)[0]
+    if isinstance(output, Sequence):
+        output_dict = process_output(dict(enumerate(output)), output_actions, rank=rank, world_size=world_size)
+        return type(output)((output_dict[i] for i in range(len(output))))
     for target, action in output_actions.items():
-        output[target] = apply_action(output[target], action, rank=rank, world_size=world_size)
+        output[target] = apply_action(output.get(target), action, rank=rank, world_size=world_size)
```
Comment (justheuristic) [inline src/tensor_parallel/slicer_wrapper.py:268]: handle optional outputs
